#Analysis of Ribose-seq Libraries
Alli Gombolay, M.P.H  
[Storici Lab] (http://www.storicilab.gatech.edu/) | [School of Biology] (http://www.biology.gatech.edu/)  
[Georgia Institute of Technology] (http://www.gatech.edu/)

Last Updated: May 2016  

##Project Overview
**Goal**: To create a streamlined bioinformatics pipeline to analyze Ribose-seq libraries on any Linux platform

**Background**: The original scripts for this analysis required LSF job scheduling software to run properly.  To enable any user to effectively and efficiently analyze their input sequencing data, I have created a streamlined, all-inclusive non-LSF dependent version of the Ribose-seq analysis pipeline.  This pipeline will run on any Linux platform.  I have also included all required input files and the appropriate directory structure that can be cloned as a GitHub repository.

**Significance**:  

**Applications**:  

##**References**  
* [Ribose-seq *Nature Methods* Paper, 2015]
(http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4686381/pdf/nihms742750.pdf)  
* [Jay Hesselberth's GitHub Page for Ribose-seq]
(https://github.com/hesselberthlab/modmap/tree/snake/pipeline/ribose-seq-ms)
* [Georgia Tech 2015 News Article on Ribose-seq]
(http://www.news.gatech.edu/2015/01/26/ribose-seq-identifies-and-locates-ribonucleotides-genomic-dna)

##Program Set-up
###Part A: Software Set-up  
* Script to download and install software automatically
* [Required Software] (https://github.com/agombolay/Ribose-seq-Project/blob/master/Required-Software.md)
* [Required Input Files] (https://github.com/agombolay/Ribose-seq-Project/blob/master/Required-Input-Files.md)

###Part B: Directory Set-up  
1. Clone the Ribose-seq Analysis Pipeline Directory Structure:  
```git clone https://github.com/agombolay/Ribose-seq-Project/tree/master/Ribose-seq-Directory```
