# Ribose-seq-Project
Alli Gombolay, M.P.H  
Storici Lab | School of Biology  
Georgia Institute of Technology  

##Project Overview
###Open-source Version of Ribose-seq Analysis Pipeline  

**References**:  
[Ribose-seq *Nature Methods* Paper, 2015]
(http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4686381/pdf/nihms742750.pdf)  
[Georgia Tech News Article on Ribose-seq]
(http://www.news.gatech.edu/2015/01/26/ribose-seq-identifies-and-locates-ribonucleotides-genomic-dna)

##Required Files:  
-Input Files: FASTQ files  
-http://amc-sandbox.ucdenver.edu/User13/outbox/2016/  

##Software Requirements:  
-bowtie  
-umitools  
-samtools  
-bedtools  
-python  
-R  
-bedToBigBed  
-bedGraphToBigWig  

##Required Directory Structure:  
1. Create a bin folder  
2. Add bin folder to PATH  
echo "export PATH="~/bin:$PATH"" >> ~/.bashrc  
3. git clone https://github.com/agombolay/Ribose-seq-Project/tree/master/Ribose-seq-Directory  

##Order of Execution of Scripts:  
1.  
2.  
3.  
4.  
