<h2><p align="center">Ribose-Map: A Bioinformatics Toolkit to Profile rNMPs Embedded in DNA</p></h2>

Alli Gombolay, M.P.H  
Contact: alli.gombolay@gatech.edu  
[Storici Lab](http://www.storicilab.gatech.edu/) | [School of Biological Sciences](http://www.biology.gatech.edu/)  
[Georgia Institute of Technology](http://www.gatech.edu/) Atlanta, GA USA

## Project Overview
**Goal**: To create a user-friendly bioinformatics toolkit to analyze ribose-seq libraries

<p align="justify">
<b>About</b>: The original scripts used to analyze ribose-seq libraries required proprietary software that is not readily available to many users, Platform Load Sharing Facility (LSF) workload management software.  To enable users to readily and efficiently analyze their ribose-seq libraries, I have created Ribose-Map, a user-friendly bioinformatics toolkit.  With minimal software dependencies and environment set-up, Ribose-Map allows users to readily profile the identity and distribution of rNMPs in the DNA of any organism of interest.  On this site, I have included extensive documentation on how to automatically download the required input files, download and install the required software, and set up the appropriate directory structure (as a GitHub repository).
</p>

## Usage
To display help menu: `script -h`  

## References  
* [Ribose-seq *Nature Methods* Paper, 2015](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4686381/pdf/nihms742750.pdf)  
* [Jay Hesselberth's GitHub Page for ribose-seq](https://github.com/hesselberthlab/modmap/tree/snake/pipeline/ribose-seq-ms)
* [Georgia Tech 2015 News Article on ribose-seq](http://www.news.gatech.edu/2015/01/26/ribose-seq-identifies-and-locates-ribonucleotides-genomic-dna)

## Program Set-up 
* How to download required input files
* How to download and install software 
* How to set up the directory structure  
```git clone https://github.com/agombolay/Ribose-seq-Project/tree/master/Ribose-seq-Directory```
